### ISSUE TITLE
Generate source maps in production? (Of course, yes!)

### ISSUE THREAD

**Issue by grahamb** [instructure/canvas-lms#1101]
I'm creating this issue as a discussion; PR forthcoming. Paging @claydiffrient & @ryankshaw to the thread.

TL; DR: sourcemaps are awesome, and they should be available in production, but node/V8 may explode trying to generate them.

Sometimes you get an error reported from a user, and it's in front-end code, and you've got to try to reproduce it and debug. Dev environments are awesome, but they're not always an option; sometimes you've got to jump in and debug it in production, or a production-like environment.

Canvas _used_ to have a neat trick where you could pass in a URL parameter and get un-minified JS. With the move to Webpack, this is gone, and as such, great sadness happens when you try to debug Canvas JS in production. I mean, have you _tried_ to step through minified, obfuscated, TinyMCE code? 😭 😭 😭

Sourcemaps are the answer to this, right? Right! They're being [generated in development](https://github.com/instructure/canvas-lms/blob/0eef2ce4b03f380e7566f52cd427e6b06630cb50/frontend_build/baseWebpackConfig.js#L75), but not in production.

Sourcemaps aren't free, though; there's a cost incurred in generating them. Specifically, two costs: time, and compute resources.

### Time
In my testing on our build machine (RHEL6 VM, quad-core Xeon 3.33GHz, 14GB RAM), generating sourcemaps took a few minutes longer than building without them:

Without sourcemaps:
```
real	5m32.714s
user	13m58.905s
sys	0m20.079s
```

With sourcemaps:
```
real	7m49.296s
user	17m35.509s
sys	0m27.662s
```

So, about 2.5 to 3 minutes longer. Not too shabby. YMMV.

### Resource Usage

Here's where things go off the rails. My first pass at this was to change `frontend_build/baseWebpackConfig` and `webpack.production.config.js` to emit sourcemaps (specifically, the `source-map` option). Now, I'm currently [about a thousand miles]( https://www.google.com/maps/dir/40.7597411,-111.8961407/Simon+Fraser+University,+8888+University+Dr,+Burnaby,+BC+V5A+1S6,+Canada/@44.9310266,-121.9362568,6z/data=!3m1!4b1!4m9!4m8!1m1!4e1!1m5!1m1!1s0x548679c0857a2227:0x506c3f6e30b55b!2m2!1d-122.9198833!2d49.2780937) away from my servers, but I imagine that it was looking something like this:

![giphy](https://user-images.githubusercontent.com/146111/29724443-c43f3420-8985-11e7-808e-86c35a58ce3f.gif)

(but a virtual dumpster file, because it's a virtual machine)

OK, but seriously, here's what happens:

```
$ CANVAS_SOURCEMAPS=youbetcha npm run webpack-production

... webpack output
... a lot of webpack output
... omg so much output

<--- Last few GCs --->

  441383 ms: Mark-sweep 1235.2 (1435.1) -> 1233.9 (1435.1) MB, 545.7 / 0.0 ms [allocation failure] [GC in old space requested].
  441957 ms: Mark-sweep 1233.9 (1435.1) -> 1233.6 (1435.1) MB, 574.0 / 0.0 ms [allocation failure] [GC in old space requested].
  442610 ms: Mark-sweep 1233.6 (1435.1) -> 1244.2 (1419.1) MB, 653.1 / 0.0 ms [last resort gc].
  443323 ms: Mark-sweep 1244.2 (1419.1) -> 1254.2 (1418.1) MB, 712.5 / 0.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x18aa27bcfb39 <JS Object>
    2: /* anonymous */ [/usr/local/atlassian/bamboo-home/xml-data/build-dir/2162691/CANVAS-BUILDDEPLOYEDGE-COMPILEASSETS/node_modules/source-map/lib/source-node.js:~80] [pc=0x1ced1dff96ac] (this=0x20d630194809 <JS Function SourceNode (SharedFunctionInfo 0x3806ba8add09)>,mapping=0x3d571e35a91 <an Object with map 0x4587d88201>)
    3: arguments adaptor frame: 3->1
    4: InnerArrayForEach(aka Inn...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [node]
 2: 0x109b7ac [node]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [node]
 5: v8::internal::Factory::NewUninitializedFixedArray(int) [node]
 6: 0xc496ff [node]
 7: v8::internal::Runtime_GrowArrayElements(int, v8::internal::Object**, v8::internal::Isolate*) [node]
 8: 0x1ced1ca079a7
sh: line 1: 20905 Aborted                 (core dumped) webpack --progress --color --config webpack.production.config.js
```

### 💣 💣 💣 💣 

_Turns out_ that webpack kinda has a problem with generating `source-maps` on Very Large Projects. Related issues: webpack/webpack#1914, webpack/webpack#2157.

The workaround for this is to allocate Node more memory, using the `--max_old_space_size` flag. Unfortunately, npm scripts that call node_module binaries don't take node flags, so to make this work while still being able to do `npm run webpack-production`, that script would need to change to `NODE_ENV=production gulp rev & node --max_old_space_size=4096 node_modules/webpack/bin/webpack.js --progress --color --config webpack.production.config.js`

> Side-note: Node 8 [adds a `NODE_OPTIONS`](https://medium.com/the-node-js-collection/node-options-has-landed-in-8-x-5fba57af703d) environment variable that would allow ` "webpack-production": "NODE_ENV=production gulp rev & NODE_OPTIONS="--max_old_space_size=4096" webpack --progress --color --config webpack.production.config.js"`. It's supposed to be back-ported to Node 6 but hasn't landed yet.

## Questions

We're going to turn on source map generation in prod here, so I'm willing to put forth a PR for it in general. It could be done in a couple of ways…

- sourcemaps on, no opt-out. Easy, just need to alter the webpack config generate them, and the npm script to pass the flag to make V8 not explode

- opt-in/out: pass an env variable (`CANVAS_SOURCEMAPS`) to `compile_assets` and down to webpack. Not too difficult, could also have two npm scripts (`webpack-production` and `webpack-production-sourceaps`)

Thoughts?

**Comment by ryankshaw** [comment]
Yep, everything you said sounds good. I'd make them on by default but piggyback off the already existing JS_BUILD_NO_UGLIFY env var that we already use and if that is set, don't do sourcemaps (that will make our jenkins CI builds run faster since either uses that).

And the max-old-space-size stuff sounds good too, we used to have to do that and then trimmed things down to where we didn't anymore but if we have to go back to making it bigger that's fine

**Comment by shawniverson** [comment]
Love the graphic @grahamb !  Uglified js debugging is a nightmare!

**Comment by grahamb** [comment]
Let's just say I had a lot of nitro cold brew before writing that issue.
